feat COS-107-109: Log entry create flow + timeline stubs

Kasia9090 opened the pull request:
**Changes**
1. feat COS-107: Created ui to log entry from organization details view
2. feat COS-108: Created rich text editor with tag support
3. feat COS-109: Enrich logger with suggested tags area at the bottom of the page
4. feat COS-142: Create ui for displaying log entry stub in the timeline


**Additionally:**
- fixed generate icon string to get viewbox from svg param (fix for icon presentation )

What types of changes does your code introduce?  _Put an `x` in the boxes that apply_

- [ ] Bugfix (non-breaking change which fixes an issue)
- [x] New feature (non-breaking change which adds functionality)
- [ ] Code style update (formatting, renaming)
- [ ] Refactoring (no functional changes, no API changes)
- [ ] Breaking change (fix or feature that would cause existing functionality to not work as expected)
- [ ] Build related changes
- [ ] Documentation Update (if none of the other choices apply)
- [ ] Other (please describe below)

## Additional context

Kasia9090 committed:
feat COS-107: Created ui to log entry from organization details view

feat COS-108: Created rich text editor with tag support

feat COS-109: Enrich logger with suggested tags area at the bottom of the page

Additionally:
- fixed generate icon string to get viewbox from svg param (fix for icon presentation )

Kasia9090 committed:
Added new SVG icons

This commit includes a set of newly added SVG icons (such as Calendar, Phone, and Mail01)

Kasia9090 committed:
Added log entry functionality and tag query to GraphQL

The commit adds two new GraphQL mutations for creating log entries and querying tags, respectively. The log entries will be useful for logging user activities while tags are necessary

Kasia9090 committed:
Ui polishes - TimelineActions.tsx component

This commit changes the color scheme for email and logger icons in the TimelineActions component based on the state. Updated icons colors to match the selected state, providing better visual cues for the current mode. Also, passed the invalidateQuery ad a prop to Logger component for better control on query invalidation.

Kasia9090 committed:
Implemented changes to the way tags are handled in the TimelineActions component. This update makes it possible to customize tags.

 Also added a new LogEntry.dto.ts file for handling the log entry data, and updated the related Logger.tsx and TagButton.tsx files to reflect these changes.

Kasia9090 committed:
(feat: COS-142): Display log entries in timeline (stubs)

Added a new component LogEntryStub.tsx that supports displaying Log Entries in the timeline. This enhances user experience by showing event related data in detail on timeline.

Updated OrganizationTimeline.tsx to incorporate LogEntryStub component. This allows timeline to handle a new event type 'LogEntry' and correctly display them using the new component.

Kasia9090 committed:
WIP

Kasia9090 committed:
Refactor TimelineActions to consolidate editor toggle logic, prevent user form accidentally exiting edit mode

Kasia9090 committed:
Disable log entry button -  this will be changed in next pr when feature is complete and delivers value

Kasia9090 committed:
Build fix

acalinica commented on packages/apps/spaces/app/organizations/[id]/components/Timeline/TimelineActions/TimelineActionsArea.tsx:
```diff
@@ -0,0 +1,28 @@
+
+  return (
+    <Box
+      bg={'#F9F9FB'}
```
Don't we have a color token for this? If not we should change this to use one.

Kasia9090 replied:
No, we dont have this color in theme

acalinica replied:
We should change it with a theme one. Let's talk with @steynviljoen67 about it.

acalinica commented on packages/apps/spaces/app/organizations/[id]/components/Timeline/TimelineActions/logger/LogEntry.dto.ts:
```diff
@@ -0,0 +1,43 @@
+export interface LogEntryDtoI {
```
Let's stick with the existing name convention

```ts
class LogEntryFormDTO extends LofEntryForm { ... }
```

Kasia9090 replied:
change applied

acalinica replied:
Can you also apply this suggestion. Naming this right is really important since other DTO's are also named in this way.

acalinica commented on packages/apps/spaces/app/organizations/[id]/components/Timeline/TimelineActions/TimelineActionsContext/TimelineActionLogEntryContext.tsx:
```diff
@@ -0,0 +1,148 @@
+  const timeoutRef = useRef<NodeJS.Timeout | null>(null);
+  const { closeEditor } = useTimelineActionContext();
+
+  const logEntryValues: LogEntryDtoI = new LogEntryDto();
```
typing this variable should not be needed since the class extends the interface.

acalinica commented on packages/apps/spaces/app/organizations/[id]/components/Timeline/TimelineActions/email/EmailTimelineAction.tsx:
```diff
@@ -0,0 +1,55 @@
+            m={6}
+            mt={2}
+            bg={'white'}
+            border='1px solid var(--gray-100, #F2F4F7)'
```
although it's perfectly right this line, I'm usually doing the following
```ts
border='1px solid'
borderColor='gray.500'
```

I do this so I can see easily read which variant of the color I'm using and to have typing.

acalinica commented on packages/apps/spaces/app/organizations/[id]/components/Timeline/TimelineActions/logger/LogEntry.dto.ts:
```diff
@@ -0,0 +1,43 @@
+  appSource: string;
+  startedAt: string | Date;
+
+  constructor(data?: any) {
```
data should not be of type any, it should be the query return type.

acalinica commented on packages/apps/spaces/app/organizations/[id]/components/Timeline/TimelineActions/logger/LogEntry.dto.ts:
```diff
@@ -0,0 +1,43 @@
+    this.startedAt = new Date().toISOString();
+  }
+
+  static toForm(data: any) {
```
Same here, data should be the query return type or at least a known entity which we have type for.

acalinica commented on packages/apps/spaces/app/organizations/[id]/components/Timeline/TimelineActions/logger/LogEntry.dto.ts:
```diff
@@ -0,0 +1,43 @@
+    return new LogEntryDto(data);
+  }
+
+  static toPayload(data: LogEntryDtoFormI) {
```
This is missing the return type which should be the Mutation input type.
```ts
static toPayload(data: LogEntryDtoFormI): LogEntryUpdateInput {
```

acalinica commented on packages/apps/spaces/app/organizations/[id]/components/Timeline/TimelineActions/logger/LogEntry.dto.ts:
```diff
@@ -0,0 +1,43 @@
+      contentType: data.contentType,
+      appSource: 'customerOS',
+      startedAt: new Date().toISOString(),
+    } as any;
```
Coercing to any is not needed if we type the return value of the function, or if you prefer coercing this object instead of typing the function, it should be `as LgoEntryUpdateInput` or however this type is called.

acalinica commented on packages/apps/spaces/app/organizations/[id]/components/Timeline/TimelineActions/logger/LogEntryTimelineAction.tsx:
```diff
@@ -0,0 +1,44 @@
+            mt={2}
+            p={6}
+            bg={'white'}
+            border='1px solid var(--gray-100, #F2F4F7)'
```
Same suggestion for this border color thing.

acalinica commented on packages/apps/spaces/app/organizations/[id]/components/Timeline/TimelineActions/logger/components/Logger.tsx:
```diff
@@ -0,0 +1,89 @@
+          tags={data?.tags}
+        />
+        <Button
+          className='customeros-remirror-submit-button'
```
why do we need a className if we style with chakra? Is this related to remirror selecting the element by classname?

acalinica commented on packages/apps/spaces/app/organizations/[id]/components/Timeline/TimelineActions/logger/components/TagSelect.tsx:
```diff
@@ -0,0 +1,190 @@
+    }
+  };
+
+  const Option = (props: any) => {
```
Can we move this in the outer scope? If not, then we should memoize it since it's going to be recreated/re-rendered on every state/prop change.

Kasia9090 replied:
it cant be memoized as it's using ref and useEffect. It also breaks when moved to outer scope. 
Left a comment in the code and will go back to it later. For now I didn't observe any performance problems

acalinica commented on packages/apps/spaces/app/organizations/[id]/components/Timeline/TimelineActions/logger/components/TagSelect.tsx:
```diff
@@ -0,0 +1,190 @@
+          {!!selectedTags?.length && (
+            <MultiCreatableSelect
+              Option={Option}
+              // autoFocus={autofocus}
```
this can be removed

acalinica commented on packages/apps/spaces/app/organizations/[id]/components/Timeline/TimelineActions/logger/components/TagSuggestor.tsx:
```diff
@@ -0,0 +1,75 @@
+              </div>
+            );
+          })}
+      </div>
```
Can we use Chakra component so that we don't need to use `cx` which basically is what chakra also uses behind to de-dupe classes and create conditional styles.

Kasia9090 replied:
it's a minor, I don't think it's worth the time now

acalinica replied:
Ok, but just so you know this should be considered debt. We should not have two ways of styling things in our project.

Kasia9090 replied:
This is suff related to external library we are using for editor

acalinica commented on packages/apps/spaces/app/organizations/[id]/components/Timeline/events/logEntry/LogEntryStub.tsx:
```diff
@@ -0,0 +1,105 @@
+      boxShadow='xs'
+      borderColor='gray.200'
+      borderRadius='lg'
+      onClick={() => console.log('opened')}
```
if this is intended ignore my comment.

acalinica commented on packages/apps/spaces/public/icons/new/message-chat-square.svg:
I'm pretty sure icons are not downloaded correctly. They have different size and already colored with gray.700 probably. They should not be downloaded from the feature design spec but from the icon library where they all have the same black color, viewbox etc. In this way we keep all icons consistent.

https://www.figma.com/file/VdWZdAh7g0RsStuBlqh3Cq/CustomerOS-Library?mode=dev

I do this by right-clicking on the icon -> Go to main component.

Kasia9090 replied:
I don't think it actually matters, as long as viewbox is consistent with path. The rest of resising is done by boxSize and other props. My change in script ensure that viewbox will always be consistent with path, no matter how the icon was downloaded

acalinica commented on packages/apps/spaces/scripts/codegen-icons.js:
```diff
@@ -1,11 +1,11 @@
 
 export const ${name} = (props: IconProps) => (
-  <Icon viewBox='0 0 24 24' fill='none' boxSize='4' {...props}>
+  <Icon viewBox='${viewBox}' fill='none' boxSize='4' {...props}>
```
please revert this as all icons in design library have and should have 24 by 24 in viewbox. If new ones are added we must enforce this 24/24 viewbox on it from the design phase.

Kasia9090 replied:
Ca you explain why? Having it hardcoded caused issues - my change resolved the issues without breaking any existing stuff.

acalinica commented on packages/apps/spaces/scripts/codegen-icons.js:
```diff
@@ -16,6 +16,11 @@ const prettierConfig = JSON.parse(
+  const match = svgString.match(/viewBox="([^"]*)"/);
+  return match ? match[1] : '0 0 24 24';  // return matched viewBox value or null if not found
+}
+
```
This is not needed as all icons must have the same viewbox. SVG artwork like grids or other graphics should be treated manually and not like icons since we cannot really reuse artworks in many places.

acalinica commented on packages/apps/spaces/styles/remirror-editor.scss:
```diff
@@ -3852,4 +3858,79 @@ a {
+  &:hover,
+  &.hovered {
+    background: var(--chakra-colors-gray-100);
+  }
```
I appreaciate the effort but we could've done this with chakra having typing all along the way. Let's leave at is for now but let's be aware in the future that we can overwrite 3rd party library styles with css-in-js where we have access to our theme.

Kasia9090 replied:
here we have access to the theme as well

acalinica commented on packages/apps/spaces/ui/form/MultiCreatableSelect/styles.ts:
```diff
@@ -0,0 +1,96 @@
+    | ChakraStylesConfig<unknown, boolean, GroupBase<unknown>>
+    | undefined,
+) => ({
+  // @ts-expect-error fixme
```
Why all the expect-error directives? Can't we fix it now?

acalinica commented on packages/apps/spaces/ui/form/RichTextEditor/TagArea.tsx:
```diff
@@ -0,0 +1,80 @@
+            const htmlValue = prosemirrorNodeToHtml(nextState?.doc);
+
+            // first update is happening before form store is initialized this change prevents error
+            if (value !== undefined) {
```
I'm aware of this and will be pushing soon a fix for it.

acalinica commented on packages/apps/spaces/ui/media/icons/EventIllLog.tsx:
```diff
@@ -0,0 +1,128 @@
+import { Icon, IconProps } from '@ui/media/Icon';
```
I would not import this as an icon looks like a heavy graphic image.

Kasia9090 replied:
this file is removed

acalinica commented on packages/apps/spaces/ui/media/icons/LeftIndent.tsx:
```diff
@@ -1,7 +1,7 @@
 
 export const LeftIndent = (props: IconProps) => (
-  <Icon viewBox='0 0 24 24' fill='none' boxSize='4' {...props}>
+  <Icon viewBox='0 0 16 16' fill='none' boxSize='4' {...props}>
```
Please re-gen icons after you revert the script script. This is potentially causing a lot of visual bugs in other places where those icons have been used.

Kasia9090 replied:
Actually it fixes the issues in other places

acalinica replied:
It might seems so but it most probably because you download an icon which has viewbox 16/16 and the script overwrites with 24/24. Then your icon looks bad but the solutions is to download the icon with vewibox 24/24, then everything will be aligned. If you need to change the viewbox of an icon which you shouldn't since they're all wrapped in a square container, you can do that by passing viewbox as props where you consume the icon.

Kasia9090 committed:
Improvements requested in the code review

Kasia9090 committed:
Update LogEntry Dto type naming

Kasia9090 committed:
Type fix

Kasia9090 committed:
Disable LogEntryDto form conversion method

Commented out the static conversion method `toForm()`. Currently, users cannot update existing data except for `startedAt` field. Hence, it is not necessary for now. This may be reactivated when user edit functionality is expanded.

Kasia9090 committed:
Remove unused 'customeros-remirror-submit-button' from Logger.tsx

The 'customeros-remirror-submit-button' was removed from Logger.tsx because it's unused.

Kasia9090 committed:
code quality improvements

Kasia9090 committed:
Added a FIXME comment to relocate the Option component

Kasia9090 committed:
Update LogEntryDto name and font weights

Renamed LogEntryDto to LogEntryFormDto to align with its specific use within form context. The change provides better clarity for future code maintenance. Font weights have also been updated in layout.tsx to match the actual font styles used. Font paths were adjusted accordingly to ensure correct font loading.

Kasia9090 committed:
Refactor interface name from LogEntryDto to LogEntryFormDto

This commit is refactoring the variable name LogEntryDto to LogEntryFormDto. The purpose of this change is to clarify that this interface is specifically for LogEntry forms, providing context for its usage. Thus, facilitating future code maintenance and understanding.
